### ISSUE TITLE
Question about `render.py`

### ISSUE THREAD

**Issue by hyoseok1223** [graphdeco-inria/gaussian-splatting#408]
Thank you for great work, I have some questions in render.py

- Below line, say frustum culled, but in this dictionary is not culled any thing, only filter by radii. frustum culling is did in differentiable rasterization right?
- In view space, i think view direction is -Z, so if do frustum culling, we only get -z_near ~ z_far? ( ex - -0.01 ~-100.0 )
- In render.py, what is viewspace_points mean? really i think it is camera space gaussian xyz point coordinate but value screenspace_points just all zero. why use viewspace_points expression? 

https://github.com/graphdeco-inria/gaussian-splatting/blob/f11001b46c5c73a0a7d553353c898efd68412abe/gaussian_renderer/__init__.py#L95C5-L100C28

```
    # Those Gaussians that were frustum culled or had a radius of 0 were not visible.
    # They will be excluded from value updates used in the splitting criteria.
    return {"render": rendered_image,
            "viewspace_points": screenspace_points,
            "visibility_filter" : radii > 0,
            "radii": radii}
```

**Comment by kwea123** [comment]
1. yes culling is done during rasterization
2. z values are in euclidean space
3. `screenspace_points` is passed to `rasterizer` function which fills this variable with the image coordinates (uv) of the gaussians

**Comment by MobiusLqm** [comment]
@kwea123 Hi, Thank your for your explanation. But Why the varible  'screenspace_points' after rendered is still filled with zero? Could you explain it more clearly?I'm still confused. Thank you!

**Comment by kwea123** [comment]
@MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86

**Comment by MobiusLqm** [comment]
> @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86

@kwea123 Greate thanks  for your explanation, I really appreciate it. I will check it later.

**Comment by ShaySheng** [comment]
> @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86

How exactly does it "retain the gradient with respect to uv"? I was under the impression that `means2D` is ultimately passed to [this section of the code](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L49). However, it appears that it isn't utilized there at all. Could you guide me on how to further trace its usage or impact?

**Comment by ShaohuaL** [comment]
> 1. yes culling is done during rasterization
> 2. z values are in euclidean space
> 3. `screenspace_points` is passed to `rasterizer` function which fills this variable with the image coordinates (uv) of the gaussians

Hi！Thank your good tutorials about 3D GS ^-^! I think means2D stores gradients about NDC space, but not image space.

`			atomicAdd(&dL_dmean2D[global_id].x, dL_dG * dG_ddelx * ddelx_dx);
			atomicAdd(&dL_dmean2D[global_id].y, dL_dG * dG_ddely * ddely_dy);
`

**Comment by JerryPW** [comment]
> > @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86
> 
> How exactly does it "retain the gradient with respect to uv"? I was under the impression that `means2D` is ultimately passed to [this section of the code](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L49). However, it appears that it isn't utilized there at all. Could you guide me on how to further trace its usage or impact?

I have the same question with you. Have you figured it out?

**Comment by MobiusLqm** [comment]
> > > @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86
> > 
> > 
> > How exactly does it "retain the gradient with respect to uv"? I was under the impression that `means2D` is ultimately passed to [this section of the code](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L49). However, it appears that it isn't utilized there at all. Could you guide me on how to further trace its usage or impact?
> 
> I have the same question with you. Have you figured it out?

sorry. I  didn't figure it out. this issue is not related with my  current project .

**Comment by Erhushenshou** [comment]
> > @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86
> 
> How exactly does it "retain the gradient with respect to uv"? I was under the impression that `means2D` is ultimately passed to [this section of the code](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L49). However, it appears that it isn't utilized there at all. Could you guide me on how to further trace its usage or impact?

I have the same problem. means2D is never used again after passing into the render function.

**Comment by FanYaning** [comment]
means2D  is hold the grad  value

> > > @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86
> > 
> > 
> > How exactly does it "retain the gradient with respect to uv"? I was under the impression that `means2D` is ultimately passed to [this section of the code](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L49). However, it appears that it isn't utilized there at all. Could you guide me on how to further trace its usage or impact?
> 
> I have the same problem. means2D is never used again after passing into the render function.


means2D is not used  in the forward function, but _RasterizeGaussians 's backward return grad values as follows:
```
        grads = (
            grad_means3D,
            grad_means2D,
            grad_sh,
            grad_colors_precomp,
            grad_opacities,
            grad_scales,
            grad_rotations,
            grad_cov3Ds_precomp,
            None,
        ) 
```
so means2D.grad=grad_means2D.

You can try this example:
```
import torch

class Exp(torch.autograd.Function):
    @staticmethod
    def forward(ctx, i, j):
        result = i.exp()
        ctx.save_for_backward(result)
        return result
 
    @staticmethod
    def backward(ctx, grad_output):
        result, = ctx.saved_tensors 
        return (grad_output * result, grad_output * result * 2)
 
if __name__ == "__main__":
    x = torch.ones(1, requires_grad = True)
    y = torch.ones(1, requires_grad = True)
    y.retain_grad()
    
    output = Exp.apply(x, y)
    print(output) # tensor([2.7183]
    
    output.backward((torch.ones(1)))
    print(y.grad)
```

**Comment by yilinshao** [comment]
> > > @MobiusLqm oh sorry, that variable is only used to keep the gradient w.r.t. uv, the content itself is still zero. The values are stored in this variable https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L265 but with this code it is very hard to extract them. You need to compute the correct offset from https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/cuda_rasterizer/rasterizer_impl.cu#L158-L165 then get it from the variable `geomBuffer` on this line https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/main/diff_gaussian_rasterization/__init__.py#L86
> > 
> > 
> > How exactly does it "retain the gradient with respect to uv"? I was under the impression that `means2D` is ultimately passed to [this section of the code](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L49). However, it appears that it isn't utilized there at all. Could you guide me on how to further trace its usage or impact?
> 
> I have the same problem. means2D is never used again after passing into the render function.

When passing `means2D` to [`_RasterizeGaussians`](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L34) during the [`forward()`](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L34) pass, it is already recorded, even though it is not explicitly used in the `forward()` function. This is because the `forward()` function is wrapped within the `apply()` method of `torch.autograd.Function`, where all inputs will be saved. Moreover, this behavior is consistent with the fact that the number of inputs to the `forward()` function should match the number of outputs in the `backward()` function, as the `backward()` method computes the gradients for all input tensors, including [`means2D`](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/diff_gaussian_rasterization/__init__.py#L145).
